File: Symmetric_Tree.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    /*
    Idea:
    In-order traverse and store the result in the vector.
    Finally find whether the elements in the vector is symmetric
    */
    bool isSymmetric(TreeNode *root) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        if (root == NULL){
            return true;
        }
        vector<int> traverseRes;
        traverseRes.clear();
        inOrderTraverse(root, traverseRes);
        int len = traverseRes.size();
        if (len % 2 == 0){
            return false;
        }
        int n;
        for (n = 0; n < len/2; n ++){
            if (traverseRes[n] != traverseRes[len - 1 - n]){
                return false;
            }
        }
        return true;
    }
    void inOrderTraverse(TreeNode *root, vector<int> &traverseRes){
        if (root == NULL) {
            return;
        }
        inOrderTraverse(root->left, traverseRes);
        traverseRes.push_back(root->val);
        inOrderTraverse(root->right, traverseRes);
    }
};

File: Go/Symmetric_Tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isMirror(r1 *TreeNode, r2 *TreeNode) bool{
    if r1 == nil && r2 == nil {
        return true
    } else if r1 != nil && r2 != nil {
        if r1.Val == r2.Val {
            if isMirror(r1.Right, r2.Left) {
                if isMirror(r1.Left, r2.Right) {
                    return true
                }
            }
        }
        return false
    } else {
        return false
    }
}
func isSymmetric(root *TreeNode) bool {
    if root == nil {
        return true
    } else {
        return isMirror(root.Left, root.Right)
    }
}
